integrationtest/record_batcher/main.d

```diff
@@ -0,0 +1,115 @@
+/*******************************************************************************
+
+    Test the behavior of RecordStream
+
+    This has to be an integration test because we test the behavior of reading
+    from stdin.
+
+    Copyright:
+        Copyright (c) 2018 sociomantic labs GmbH. All rights reserved
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module integrationtest.record_batcher.main;
+
+import core.sys.posix.unistd;
+
+import ocean.core.Test;
+import ocean.io.Console;
+import ocean.io.device.Device;
+import ocean.transition;
+
+import swarm.util.RecordStream;
+
+import ocean.io.stream.Buffered;
+import core.sys.posix.unistd;
+import core.sys.posix.sys.wait;
+
+
+version (UnitTest) {} else
+public int main ()
+{
+    int[2] pipes;
+    assert(pipe(pipes) == 0);
+
+    if (auto pid = fork())
+    {
+        return runSender(pid, pipes[1]);
+    }
+    else
+    {
+        dup2(pipes[0], STDIN_FILENO);
+        runListener();
+        return 0;
+    }
+}
+
+public int runSender (pid_t pid, int write_fd)
+{
+    mstring buffer;
+    // We need the destructor to be called at the end, to close the fd
+    {
+        scope conduit = new FDConduit(write_fd);
+        foreach (record; Input)
+            record.serialize(conduit, buffer);
+
+        sleep(2);
+        // const istring msg = "Hello world!";
+        // write(write_fd, msg.ptr, msg.length);
+        //close(write_fd);
+    }
+
+    int status;
+    auto ret = waitpid(pid, &status, 0);
+    assert(ret > 0, "waitpid returned <= 0");
+    return WIFEXITED(status) ? WEXITSTATUS(status) : -1;
+}
+
+public void runListener ()
+{
+    size_t index;
+    scope comparer = (Record r)
+        {
+            test!("==")(r.key, Output[index].key);
+            test!("==")(r.value, Output[index].value);
+            index++;
+            return index < Output.length;
+
+        };
+    scope stream = new StdinRecordStream(comparer);
+    auto eos = stream.process();
+    assert(!eos, "End Of Stream reached!");
+}
+
+
+private const Input = [
+    Record(null,     [ 0x52, 0x45, 0x50, 0x4F, 0x52, 0x54, 0x49, 0x4E, 0x47 ]),
+    Record(null,     [ 0x20, 0x57, 0x41, 0x53 ]),
+    // Keys are 16 hex char
+    Record([ 0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x42, 0x42,
+             0x42, 0x42, 0x42, 0x42, 0x42 ],
+           [ 0x48, 0x65, 0x6C, 0x6C, 0x6F, 0x20, 0x57, 0x6F, 0x72, 0x6C, 0x64 ]),
+];
+
+private const Output = [
+    Input[0],
```

Comment:
Why is the receiver not comparing the received records against the same array as the sender is sending? (Also why one less element?)

Reply:
For the story, what happened is that I originally tested ONLY corner cases and was seeing weird results.
Basically if you `serialize` a record, it doesn't check if  they key is either `null` / empty or if it's 16 chars. So if you put `Record([ 0x20 ], [ ... ])` it will gladly serialize it, and will screw up the deserialization big time. I will try to implement a fix for this in v5.x.x after this is merged.

That being said - I will remove the `Output` array.